PHP: Make the default package more sane [v3]

etu opened the pull request:
###### Motivation for this change
Our old PHP derivation was crazy big with a crazy amount of enabled extensions that also had a crazy amount of defaults that we pulled in, yet excluded some extensions that would have been nice to have by default.

This is a follow up on: #82348 to utilize #79377 more. And also a follow-up on all of the work done by me and @talyz in #82794 and #83514

But this grew a bit more, and was more complex than I imagined when I started this. But it became a collaboration project with @talyz to finish everything up.

We have also written the initial documentation for PHP in NixOS.

#### The current PHP
The current closure size is:
```
/nix/store/4ga497cqvz04vcyw9lf7s74ymrc6r28g-php-7.4.3	234.3M
````

And this contains things (among others) like:
```
/nix/store/y5166rf2lkxkf3991qm390vn92czgfgh-openldap-2.4.49          	 53.6M
/nix/store/2zsgfrkp0hn878rnrp0azpmy76ljm2b0-linux-pam-1.3.1          	 53.6M
```

#### The new smaller PHP
Here we have a smaller closure size:
```
/nix/store/10867i6grp8w8p6p6k1w8rxiwza5nyyy-php-with-extensions-7.4.3	189.4M
```

And we have a smaller base package:
```
/nix/store/v5v6wn207ibpp6n90z05f11k7svq94wd-php-7.4.3	 85.5M
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

etu committed:
php: Make a slimmer php as default

etu committed:
phpPackages.exts: Drop uneeded extensions

etu committed:
php: Add release log entry for the php changes

etu committed:
php-embed: Drop the declaration of the php-embed attributes

etu committed:
php-unit: Drop the declaration of the php-unit attributes since they aren't used

etu committed:
doc/php: Add initial documentation for PHP

etu committed:
php: init phpXYbase packages and make the default php extended by default

Also, add opcache to default extensions since it significantly
increases PHP's performance and is by default enabled on Debian based
distributions. Not having it enabled by default results in a puzzling
performance loss for anyone attempting to migrate from Debian/Ubuntu
to NixOS who is unaware of this. Therefore, enable it by default. /talyz

etu committed:
phpPackages.exts.gmp: Enable on darwin

etu committed:
unit: Make unit use phpbase packages

talyz committed:
php: Build an even slimmer base

This moves yet more extensions from the base build to
phpPackages.ext. Some of the extensions are a bit quirky and need
patching for this to work, most notably mysqlnd and opcache.

Two new parameters are introduced for mkExtension - internalDeps and
postPhpize. internalDeps is used to specify which other internal
extensions the current extension depends on, in order to provide them
at build time. postPhpize is for when patches and quirks need to be
applied after running phpize.

Patch notes:

- For opcache, older versions of PHP have a bug where header files are
  included in the wrong order.

- For mysqlnd, the config.h is never included, so we include it in the
  main header file, mysqlnd.h. Also, the configure script doesn't add
  the necessary library link flags, so we add them to the variable
  configure should have added them to.

talyz committed:
phpPackages.exts: Run tests by default

The tests for many of the extensions run just fine, for some a small
portion fail. This runs the tests by default and disables the tests
extensions with any failing tests.

talyz committed:
php: Make buildEnv recursive + take extension deps into account

A slight rewrite of buildEnv which:

1. Makes buildEnv recursively add itself to its output, so that it can
   be accessed from any php derivation.

2. Orders the extension text strings according to their internalDeps
   attribute - dependencies have to be put before dependants in the
   php.ini or they will fail to load due to missing symbols.

etu committed:
docs/rl: Update release log to match the updates to the default attributes

talyz committed:
php: Make mkExtension pass on additional args to mkDerivation

etu committed:
php: Wrap the programs and provide an environment variable instead of a flag

talyz committed:
phpPackages: Move phpPackages to php.packages

This means php packages can now refer to other php packages by looking
them up in the php.packages attribute and gets rid of the internal
recursive set previously defined in php-packages.nix. This also means
that in applications where previously both the php package and the
corresponding version of the phpPackages package set had to be
specified, the php package will now suffice.

This also adds the phpWithExtensions parameter to the
php-packages.nix, which can be used by extensions that need a fully
featured PHP executable.

talyz committed:
php.packages: Use derivations for internalDeps in mkExtension

Make mkExtension put headers in the dev output and use them, instead of
a different part of the current source tree, when referring to another
extension by using internalDeps.

This means external extensions can be built against the internal ones.

talyz committed:
php.packages.buildPecl: Introduce built-in extension dependencies

Introduce the internalDeps attribute used by mkExtension, to refer to
internal php extensions built by mkExtension.

talyz committed:
php.buildEnv: Put the extraConfig snippet after extensions

Make sure all the extensions are loaded first, then read the custom
config where possible.

talyz committed:
php.packages: Fix and clean up external extensions

Fix extensions that were broken by the extension refactoring and
remove pthreads and pinba, which had asserts which broke evaluation,
were in need of refactoring and of dubious value.

talyz committed:
nixos/nextcloud: Get nextcloud working

etu committed:
nixos/nextcloud: upgrade to php74

etu committed:
nixos/php: Add tests for phpfpm

This test checks that we evaluate PHP properly and that certain
extensions are actually loaded.

etu committed:
nixos/php: Move the pcre tests to the php test attribute

etu commented:
@GrahamcOfBorg build php72 php73 php74 php72Packages.exts php73Packages.exts php74Packages.exts nixosTests.nextcloud nixosTests.php

aanderse reviewed (commented):
I haven't been through all of this yet, but here are a few initial suggestions for wording in the documentation if anyone find them useful.

Izorkin commented on pkgs/top-level/php-packages.nix:
```diff
@@ -592,41 +583,17 @@ let
-
-    buildInputs = with pkgs; [ pcre2.dev ];
-  };
-
```
Pthreads and pinba phpPackages removed?

etu replied:
For now yes because they did not build. Whenever this change is pulled in and build it's easier to add things back and actually test it in a better way.

We reasoned that we had a set of changes that was big enough as it was.

So for now I think that's best, but I won't reject adding it back in a better way if they build.

Izorkin replied:
This pakages builded with this patch

```
diff --git a/pkgs/top-level/php-packages.nix b/pkgs/top-level/php-packages.nix
index 21207c72fe9..7dd422e839d 100644
--- a/pkgs/top-level/php-packages.nix
+++ b/pkgs/top-level/php-packages.nix
@@ -18,6 +18,7 @@ let
     pname = "php-${pname}";
   });

+  isPhp72 = pkgs.lib.versionAtLeast php.version "7.2";
   isPhp73 = pkgs.lib.versionAtLeast php.version "7.3";
   isPhp74 = pkgs.lib.versionAtLeast php.version "7.4";

@@ -295,7 +296,7 @@ in
   # This is a set of PHP extensions meant to be used in php.buildEnv
   # or php.withExtensions to extend the functionality of the PHP
   # interpreter.
-  extensions = {
+  extensions = rec {
     apcu = buildPecl {
       version = "5.1.18";
       pname = "apcu";
@@ -578,6 +579,53 @@ in
       configureFlags = [ "--with-excel" "--with-libxl-incdir=${pkgs.libxl}/include_c" "--with-libxl-libdir=${pkgs.libxl}/lib" ];
     };

+    pinba = if isPhp73 then pinba112 else pinba111;
+
+    pinba111 = buildPecl {
+      version = "1.1.1";
+      pname = "pinba";
+
+      src = pkgs.fetchFromGitHub {
+        owner = "tony2001";
+        repo = "pinba_extension";
+        rev = "RELEASE_1_1_1";
+        sha256 = "1kdp7vav0y315695vhm3xifgsh6h6y6pny70xw3iai461n58khj5";
+      };
+
+      meta = with pkgs.lib; {
+        description = "PHP extension for Pinba";
+        longDescription = ''
+          Pinba is a MySQL storage engine that acts as a realtime monitoring and
+          statistics server for PHP using MySQL as a read-only interface.
+        '';
+        homepage = "http://pinba.org/";
+        broken = (isPhp73);
+      };
+    };
+
+    pinba112 = buildPecl {
+      version = "1.1.2-dev";
+      pname = "pinba";
+
+      src = pkgs.fetchFromGitHub {
+        owner = "tony2001";
+        repo = "pinba_extension";
+        rev = "edbc313f1b4fb8407bf7d5acf63fbb0359c7fb2e";
+        sha256 = "02sljqm6griw8ccqavl23f7w1hp2zflcv24lpf00k6pyrn9cwx80";
+      };
+
+      meta = with pkgs.lib; {
+        description = "PHP extension for Pinba";
+        longDescription = ''
+          Pinba is a MySQL storage engine that acts as a realtime monitoring and
+          statistics server for PHP using MySQL as a read-only interface.
+        '';
+        homepage = "http://pinba.org/";
+
+        broken = (!isPhp73);
+      };
+    };
+
     protobuf = buildPecl {
       version = "3.11.2";
       pname = "protobuf";
@@ -595,6 +643,40 @@ in
       };
     };

+    pthreads = if isPhp73 then pthreads32-dev else pthreads32;
+
+    pthreads32 = buildPecl rec {
+      version = "3.2.0";
+      pname = "pthreads";
+
+      src = pkgs.fetchFromGitHub {
+        owner = "krakjoe";
+        repo = "pthreads";
+        rev = "v${version}";
+        sha256 = "17hypm75d4w7lvz96jb7s0s87018yzmmap0l125d5fd7abnhzfvv";
+      };
+
+      buildInputs = with pkgs; [ pcre.dev ];
+
+      meta.broken = (!isPhp72 || isPhp73);
+    };
+
+    pthreads32-dev = buildPecl {
+      version = "3.2.0-dev";
+      pname = "pthreads";
+
+      src = pkgs.fetchFromGitHub {
+        owner = "krakjoe";
+        repo = "pthreads";
+        rev = "4d1c2483ceb459ea4284db4eb06646d5715e7154";
+        sha256 = "07kdxypy0bgggrfav2h1ccbv67lllbvpa3s3zsaqci0gq4fyi830";
+      };
+
+      buildInputs = with pkgs; [ pcre2.dev ];
+
+      meta.broken = (!isPhp73 || isPhp74);
+    };
+
     redis = buildPecl {
       version = "5.1.1";
       pname = "redis";

```

etu replied:
Pinba is now added back in 1e42497447d385a87a3c8df21f269575d2061c3a.

A bit differently written but it's there. I will work on pthreads next.

etu replied:
Pthreads builds now for me on php72 and php73 but not php74. And it is archived by upstream: https://github.com/krakjoe/pthreads

So it doesn't seem to be supported or updated anymore. So I'll drop that one for now unless you have a patch for it that works on all versions.

etu replied:
Ok, added it as dccccab1e24 and marked it as broken for php74. Now it's back to how it was before :)

Izorkin replied:
> Pthreads builds now for me on php72 and php73 but not php74. And it is archived by upstream: https://github.com/krakjoe/pthreads

On php 7.4 package pthreads replaced to https://github.com/krakjoe/parallel

flokli replied:
I propose to do this in a followup PR. No need to delay this PR any further :-)

etu replied:
Yeah, not adding anything new here. But I restored previous state at least :-)

jtojnar commented on doc/languages-frameworks/php.section.md:
```diff
@@ -0,0 +1,72 @@
+We keep releases as long as they are supported for the entirety of the next
+NixOS release life cycle. See [PHP Supported Versions](https://www.php.net/supported-versions.php).
+
+As for packages we have `php.phpPackages` that contains a bunch of attributes where
```
`php.phpPackages` feels too verbose. Could we just do `php.packages` or `php.pkgs` like Python does?

talyz replied:
Good idea! I'll fix that right away.

talyz replied:
Done!

jtojnar commented on doc/languages-frameworks/php.section.md:
````diff
@@ -0,0 +1,72 @@
+
+```nix
+let
+  myPhp = php.buildEnv { exts = pp: with pp; [ imagick exts.opcache ]; };
````
Would it make sense to mirror other language infrastructures like Python, which use `withPackages`?

There is also https://discourse.nixos.org/t/wip-rfc-a-new-nixpkgs-frontend-for-language-infrastructures/3447

etu replied:
It would, `php.buildEnv` does more things than just adding packages though. And it's something that already exists in `master`. So just renaming it wouldn't make sense. But we can probably add a wrapper for `buildEnv` named `withPackages`. But it's not a super big thing and could be done later on as well.

aanderse replied:
I thought `withPackages` was explicitly rejected by someone... (@FRidh maybe) because that would be misleading. I'm having trouble revealing the details at the moment.

FRidh replied:
My remark was that `withPackages` should only take a lambda, not an attribute set
https://github.com/NixOS/nixpkgs/pull/79377#issuecomment-583104499

aanderse replied:
@FRidh ah, thanks for clarifying!

etu replied:
I think a `withPackages` wrapper can be added later on after this PR as well. It's a quite small thing to add to make it simpler to use. But for now I'm fairly happy with all the progress we have here.

We have a bunch of small things we would like to improve after this PR. But we had enough changes already and tried to stick to "a minimal" amount of changes.

We (me and @talyz) were discussing how we possibly could break it down into several PRs and came to the conclusion that we can't.

talyz commented on doc/languages-frameworks/php.section.md:
```diff
@@ -0,0 +1,72 @@
+
+#### Overview
+
+Several versions of PHP is available on Nix, as well as a bunch of extensions
```
I've addressed all of these and added some additional fixes, so please have another look :)

talyz commented:
@GrahamcOfBorg build php72 php73 php74 php72Packages.exts php73Packages.exts php74Packages.exts nixosTests.nextcloud nixosTests.php

aanderse commented:
My only remaining comment is that I wonder if it becomes confusing (or overly verbose) to have `php.packages.exts`, especially for people newer to nixpkgs. Currently in my mind I wonder if `php.packages` and `php.extensions` (or `php.pkgs` and `php.exts`) is less confusing, but of course that would potentially end up making `buildEnv` more verbose. To alleviate that maybe we could have `buildEnv` operate on `php` instead of `pp`, like `php.buildEnv { exts = with p; [ pkgs.imagick exts.opcache ]; }`... but maybe I'm just rambling at this point and this doesn't make any sense... feel free to disregard :smile:

talyz commented:
@aanderse Yes, that's also something we discussed, but decided that the scope of the PR was big enough as is. Definitely something to think about, though :)

flokli commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9371,86 +9371,15 @@ in
-
-  phpPackages-unit = php74Packages-unit;
 
-  php72Packages-unit = recurseIntoAttrs (callPackage ./php-packages.nix {
```
I very much approve inlining all these into the only derivations that use it, but we still need to add the attributes to `pkgs/top-level/aliases.nix` with a throw explaining they were inlined, and how to obtain a similarily-configured PHP.

etu replied:
I've now made two fixup commits to add these to the aliases files, the commit ids are 9e51ba64d9e and 98b77d59ba0.

Me and @talyz can squash them in later :-)

flokli reviewed (commented):
Only read over the docs and skimmed over the code so far, haven't actually run anything yet.

FRidh commented:
> Currently in my mind I wonder if php.packages and php.extensions (or php.pkgs and php.exts) is less confusing, but of course that would potentially end up making buildEnv more verbose.

I think we should typically go for the verbose options. Readability is important. Now, we already use `pkgs` everywhere, so maybe not change that one, but `exts` is...strange.

etu commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -125,6 +125,71 @@
    </listitem>
+   <listitem>
+     <para>
+       Since this release we have an easy way to customize your PHP install to get a much smaller
```
Fixed :)

etu commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -125,6 +125,71 @@
+   <listitem>
+     <para>
+       Since this release we have an easy way to customize your PHP install to get a much smaller
+       base PHP with only wanted extensions enabled. See following snippet to install a smaller PHP
```
Fixed :-)

etu commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -125,6 +125,71 @@
+     <para>
+       Since this release we have an easy way to customize your PHP install to get a much smaller
+       base PHP with only wanted extensions enabled. See following snippet to install a smaller PHP
+       with <literal>imagick</literal>, <literal>opcache</literal> and <literal>pdo_mysql</literal>:
```
I've clarified on this now :-)

etu commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -125,6 +125,71 @@
+];</programlisting>
+
+       The default <literal>php</literal> attribute haven't lost any extensions by default though. It
+       has added the <literal>opcache</literal> extension as an additional extension.
```
Fixed as well :-)

etu commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -125,6 +125,71 @@
+     <para>
+       Since we have a smaller base package that we base the main <literal>php</literal> on a
+       smaller base package we've decided to remove a big bunch of options to make the main
+       PHP derivation much easier to work with.
```
I've re-written this to make more sense.

The project to do this has been big we've gone back and forth on a bunch of things a couple of times while working on it. It has resulted in things (especially descriptions) that may be outdated here and there. :-)

etu commented:
> > Currently in my mind I wonder if php.packages and php.extensions (or php.pkgs and php.exts) is less confusing, but of course that would potentially end up making buildEnv more verbose.
> 
> I think we should typically go for the verbose options. Readability is important. Now, we already use `pkgs` everywhere, so maybe not change that one, but `exts` is...strange.

Yeah, but since there's a lot of extensions from upstream and a smaller set of extensions from 3rd party sources. We kinda wanted to make a distinction between the two types since that can indicate how reliable and maintained they are.

etu commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -125,6 +125,71 @@
+
+       <itemizedlist>
+         <title>PHP <literal>config</literal> flags that we don't read anymore:</title>
+         <listitem><para><literal>config.php.argon2</literal></para></listitem>
```
They are needed, I've just tried to remove the `para` block and then the manual doesn't build anymore :-)

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -6,27 +6,27 @@ let
-    ${optionalString cfg.caching.memcached "extension=${phpPackages.memcached}/lib/php/extensions/memcached.so"}
-    extension=${phpPackages.imagick}/lib/php/extensions/imagick.so
-    zend_extension = opcache.so
-    opcache.enable = 1
```
Is `opcache.enable = 1` not needed anymore? This isn't part of `extraConfig` anymore AFAICS.

talyz replied:
Opcache is enabled by default if the module is loaded, according to the [PHP website](https://www.php.net/manual/en/opcache.configuration.php). I'll check if this is the case when running the binary.

talyz replied:
It is indeed on by default.

talyz commented:
> > Currently in my mind I wonder if php.packages and php.extensions (or php.pkgs and php.exts) is less confusing, but of course that would potentially end up making buildEnv more verbose.
> 
> I think we should typically go for the verbose options. Readability is important. Now, we already use `pkgs` everywhere, so maybe not change that one, but `exts` is...strange.

I agree - this should be changed. Since we now have `php.packages` I don't see why we wouldn't move `php.packages.exts` to `php.extensions`. However, just to be clear: `exts` was not introduced in this PR - it only moves `phpPackages`, which `exts` was already a part of, to `php.packages`.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -6,27 +6,27 @@ let
+      pp.redis pp.memcached pp.imagick
+    ];
+    extraConfig = phpOptionsStr;
+  };
```
Have you confirmed that this config is actually taken into account by phpfpm entirely? The module adds another `-c php.ini` in the systemd-unit which IIRC overrides changes from `php.buildEnv`:

https://github.com/NixOS/nixpkgs/blob/24d456a48daccf1f8b179c53d92ed3c2c3bc47b4/nixos/modules/services/web-servers/phpfpm/default.nix#L276

talyz replied:
That's why `php.buildEnv` was rewritten to not use the `-c`-flag. It instead uses an environment variable to set additional config.

Ma27 commented on pkgs/top-level/php-packages.nix:
```diff
@@ -478,50 +513,6 @@ let
 
-  pinba = if isPhp73 then pinba73 else pinba7;
-
-  pinba7 = assert !isPhp73; buildPecl {
```
Are we still using this approach somewhere to ensure that an extension is only built against supported PHP versions? If yes, we can probably do some sort of `disabled = <cond>` (as it's done in python packages) IMHO.

talyz replied:
I got rid of all uses in `php-packages.nix` by replacing them with `meta.broken` or, in the case of `pinba` and `pthreads`, removing them. A `disabled` param would be nice, though :)

etu commented:
> > > Currently in my mind I wonder if php.packages and php.extensions (or php.pkgs and php.exts) is less confusing, but of course that would potentially end up making buildEnv more verbose.
> > 
> > 
> > I think we should typically go for the verbose options. Readability is important. Now, we already use `pkgs` everywhere, so maybe not change that one, but `exts` is...strange.
> 
> I agree - this should be changed. Since we now have `php.packages` I don't see why we wouldn't move `php.packages.exts` to `php.extensions`. However, just to be clear: `exts` was not introduced in this PR - it only moves `phpPackages`, which `exts` was already a part of, to `php.packages`.

It would be nice to separate extensions from command line utilities... It would make it much nicer :D

etu committed:
php: split php.packages to php.packages and php.extensions

So now we have only packages for human interaction in php.packages and
only extensions in php.extensions. With this php.packages.exts have
been merged into the same attribute set as all the other extensions to
make it flat and nice.

The nextcloud module have been updated to reflect this change as well
as the documentation.

etu commented:
@GrahamcOfBorg eval

etu committed:
php: passthru .dev output as well

etu commented:
@GrahamcOfBorg build php72 php73 php74 php72.extensions php73.extensions php74.extensions php72.packages php73.packages php74.packages nixosTests.nextcloud nixosTests.php unit uwsgi

talyz committed:
php: Simplify php-packages import, rename exts -> extensions

talyz committed:
php: Add withExtensions as a simpler alternative to buildEnv

talyz committed:
php: Add missing hash extension to php < 7.4

talyz committed:
php: Document withExtensions + general improvements

etu commented:
@GrahamcOfBorg eval

talyz commented:
@jtojnar @aanderse I've now added a `php.withExtensions` to provide a simpler interface when the additional config option provided by `php.buildEnv` is unnecessary.

I also renamed exts to extensions in `php.buildEnv` and rewrote some of the documentation.

Me and @etu now feel like this is done and that all raised concerns have been addressed.

Ma27 reviewed (approved):
Awesome work :+1:

flokli commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -6,27 +6,27 @@ let
-  phpPackages = pkgs.php73Packages;
+  phpPackage = pkgs.php74.buildEnv {
+    exts = pp: with pp; [
+      bcmath calendar curl exif ftp filter gd gettext gmp intl json ldap
+      mysqlnd opcache openssl pcntl pdo pdo_mysql pdo_odbc pdo_pgsql
+      pdo_sqlite pgsql readline session soap sodium sqlite3 zip zlib mbstring
+      posix ctype dom simplexml xmlreader xmlwriter
+      apcu redis memcached imagick
```
I'd assume these are the default extensions, plus the additional `acpu`, `redis`, `memcached` and `imagick` extensions - or did you go through the requirement list of nextcloud?

If it's just the default extensions, plus some additional ones, could we expose the default ones, maybe via `pkgs.php74.defaultExtensions`? That'd make things way more obvious.

I know it's less explicit, but should

etu replied:
You're exactly right. It's the default list plus apcu, redis, memcached and imagick.

After we get this in I could make that available and document how to extend PHP based on the default set. I think our changeset is big enough as it is :-)

etu replied:
@flokli How would you like a `php.withAdditionalExtensions` or `php.withExtraExtensions` that takes the default list and adds extra ones? :)

Of course we could expose the default list as well.

flokli replied:
Disclaimer: This disucssion should not block the PR. It's good to go IMHO!

Exposing the list of default extensions is probably already by itself quite readable and enough. A `withAdditionalExtensions` doesn't add too much benefit IMHO, and wouldn't allow me to do things like: take all the default extensions, but remove module x, and replace it with a patched version of it…

talyz replied:
I've added an `enabledExtensions` attribute and documentation about its use. I think this should be passed as an additional argument to `withExtensions` and `buildEnv.extensions`, optimally, but that would need much more work and this should suffice for now.

flokli reviewed (approved):
LGTM! One minor UX question, but no reason to block this PR.

aanderse reviewed (approved):
I played around with this today and it worked well :+1:

@jtojnar are you good with this?
@etu @talyz thanks for being superstars!

jtojnar commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -125,6 +125,73 @@
+
+       <programlisting>
+environment.systemPackages = [
+(pkgs.php.buildEnv { exts = pp: with pp; [
```
```suggestion
(pkgs.php.buildEnv { extenssions = pp: with pp; [
```

etu replied:
Thanks! Fixed both! :)

jtojnar commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -6,27 +6,27 @@ let
-  phpPackage = pkgs.php73;
-  phpPackages = pkgs.php73Packages;
+  phpPackage = pkgs.php74.buildEnv {
+    exts = pp: with pp; [
```
```suggestion
    extensions = pp: with pp; [
```

etu commented:
Since we have approval from several people that care about this ecosystem and addressed the concerns by people that care about similar ecosystems within nixpkgs I think we can call this good to merge as it is.

So if nobody speaks up before 20:00 (CEST) today (that is in 8:25 from the time of this posting) unless someone speaks up before then or just merge it themself before that :-)

etu committed:
php.extensions.pinba: Reintroduce pinba for php72, php73 and php74

etu committed:
php.extensions.pthreads: Reintroduce pthreads for php72 and php73

talyz committed:
php: Add enabledExtensions attribute to PHP derivations

This provides a means to build a PHP package based on a list of
extensions from another.

For example, to generate a package with all default extensions
enabled, except opcache, but with ImageMagick:

php.withExtensions (e:
  (lib.filter (e: e != php.extensions.opcache) php.enabledExtensions)
  ++ [ e.imagick ])

talyz committed:
nixos/nextcloud: Use php.enabledExtensions

talyz commented:
I've added an `enabledExtensions` attribute and documentation about its use, fixed a typo in the `nextcloud` module which broke evaluation and fixed a few other minor things. The nextcloud and php tests pass.

aanderse commented:
> I've added an `enabledExtensions` attribute and documentation about its use, fixed a typo in the `nextcloud` module which broke evaluation and fixed a few other minor things. The nextcloud and php tests pass.

ping @FRidh @jtojnar @etu @Ma27 @flokli does `enabledExtensions` sound good to you?

Izorkin commented on pkgs/development/interpreters/php/default.nix:
```diff
@@ -204,8 +101,6 @@ let
           --replace '@PHP_LDFLAGS@' ""
       done
 
-      substituteInPlace ./build/libtool.m4 --replace /usr/bin/file ${file}/bin/file
```
This replace don't need?
Error - `./configure: line 84378: /usr/bin/file: No such file or directory`

etu replied:
Do you find anything that is broken due to removing this line other than this warning?

Having this line in adds `5.7M` to the closure size, and if that only is to get rid of a warning but actually not solve any issues, then we're not pulling file in.

Because 5.7M is 6% added closure size.

Izorkin replied:
ok

flokli commented on pkgs/development/interpreters/php/default.nix:
```diff
@@ -243,57 +144,121 @@ let
+              inherit buildEnv withExtensions enabledExtensions;
+              inherit (php-packages) packages extensions;
+            };
+            paths = [ php ];
```
@etu It seems this currently causes the phpBase  to be considered (and used) when patching shebangs in CLI tools - see `${arcanist}/bin/arcanist --help`.

flokli replied:
See for an example before and after: https://pastebin.com/JX1N07dv

etu replied:
I'm thinking that this must be due to `symlinkJoin` that was introduced in https://github.com/NixOS/nixpkgs/pull/79377.

But the problem became visible now because of the slimmer base.

We also have some weird issue with `nix-shell -p php74`, we get `php74base` in reality. And it can be confirmed by looking into the `$PATH` of the shell, we don't get the `php-with-extensions` derivation that we would expect.

flokli replied:
@Ma27, @aanderse , could you take a look?

talyz replied:
It's because we inherit the `dev` output from the original php package. I'm working on a fix right now.

talyz replied:
I've opened #84721 for the fix.

chpio commented:
hi, i have a custom extension, that i need for my script. The extension build seems to work, but it's not available in `php.packages`. Im relatively new to nix, so pardon me.

my nix shell:

```
let
	pkgs = import <nixpkgs> {};
	phpGeos = pkgs.stdenv.mkDerivation rec {
		pname = "php-geos";
		version = "1.0.0";
		meta = {
			description = "php binding for GEOS";
		};

		src = pkgs.fetchurl {
			url = "https://git.osgeo.org/gitea/geos/php-geos/archive/${version}.tar.gz";
			sha256 = "0d90z591s092ys0d9gww2xw8vsvcdl6jbc90cgc6avq27dx4xk89";
		};

		nativeBuildInputs = [ pkgs.autoconf pkgs.automake pkgs.php74base ];
		preConfigure = "phpize";

		buildInputs = [ pkgs.geos ];

		installPhase = ''
			mkdir -p $out/lib/php/extensions
			cp modules/geos.so $out/lib/php/extensions/geos.so
		'';
	};
	php = pkgs.php74.buildEnv {
		extensions = exts: pkgs.php74.enabledExtensions ++ [phpGeos];
	};
in
	pkgs.stdenv.mkDerivation {
		name = "...";
		buildInputs = [
			php
			php.packages.composer
		];
	}

```

i have assumed that all enabled extensions would be passed to packages. When i run `php -i` my ext shows up. But when running `/nix/store/pdllmyi9bdk7hc0bxgr8x01z27wgrdi9-php-with-extensions-7.4.4/bin/php -i` (the php bin that composer is run with) the ext is not listed.

etu commented:
Hi @chpio!

You're about right all the way there!

Composer get's it's own environments PHP, and currently on master it doesn't use the overriden php. But this should be totally doable when we get #85026 together.

With that PR you should be able to use `customPhp.packages.composer` for instance.

This isn't something that have been possible at all before in Nix so this will be a very exciting development :-)

If you can it would be great if you can check out that PR and report back there if it works for you.

@talyz and me will probably team up and try to get that one shaped up some day as well so we can get it in.

If you want a faster workaround in the meanwhile you could make your own derivation for composer that just use your custom php since composer is only a phar anyways.

Something like this can probably work for your case:
```nix
let
  myphp = pkgs.php74; # customize here
  mycomposer = writeScriptBin "composer" ''
    #!/bin/sh
    ${myphp} ${myphp.packages.composer}/libexec/composer/composer.phar \"$@\"
  '';
in {}
```

The only thing I *know* is missing from that composer is to inject unzip into the `$PATH` in that script.

talyz commented:
Hi @chpio! Can you try #85026 now? Your use case should work fine now :)
